libeventheader-decode-cpp/src/EventFormatter.cpp

```diff
@@ -782,48 +882,206 @@ WriteUcsVal(
     assert(9u <= sb.Room());
     if (json)
     {
-        sb.WriteUtf8Byte('"');
+        sb.WriteUtf8ByteUnchecked('"');
         sb.WriteUcsCharJsonEscaped(ucs4); // UCS1
-        sb.WriteUtf8Byte('"');
+        sb.WriteUtf8ByteUnchecked('"');
     }
     else
     {
         sb.WriteUcsChar(ucs4); // UCS1
     }
 }
 
+// Requires: there is room for 18 chars.
+static void
+WriteFloat32(
+    StringBuilder& sb,
+    uint32_t valSwapped,
+    bool json)
+{
+    float valFloat;
+    static_assert(sizeof(valFloat) == sizeof(valSwapped), "Expected 32-bit float");
+    memcpy(&valFloat, &valSwapped, sizeof(valSwapped));
+    bool const needQuote = json && !isfinite(valFloat);
+    sb.WriteQuoteIf(needQuote);
+    sb.WriteNumber(18 - 2, valFloat);
+    sb.WriteQuoteIf(needQuote);
+}
+
+// Requires: there is room for 27 chars.
 static void
-AppendUtf8(
+WriteFloat64(
+    StringBuilder& sb,
+    uint64_t valSwapped,
+    bool json)
+{
+    double valFloat;
+    static_assert(sizeof(valFloat) == sizeof(valSwapped), "Expected 64-bit double");
+    memcpy(&valFloat, &valSwapped, sizeof(valSwapped));
+    bool const needQuote = json && !isfinite(valFloat);
+    sb.WriteQuoteIf(needQuote);
+    sb.WriteNumber(27 - 2, valFloat);
+    sb.WriteQuoteIf(needQuote);
+}
+
+static void
+AppendUtf8Unchecked(
     StringBuilder& sb,
     std::string_view utf8)
 {
     sb.EnsureRoom(utf8.size());
-    sb.WriteUtf8(utf8);
+    sb.WriteUtf8Unchecked(utf8);
 }
 
+// Requires: utf8.size() <= roomReserved <= sb.Room().
+// Postcondition: sb.Room() >= roomReserved - utf8.size(), i.e. maintains "extra" space.
+//
+// If Json is true, performs Json escape.
+//
+// Unicode (non)conformance:
+// - Accepts (passes-through) 3-byte sequences that decode to code points in the
+//   surrogate range.
+// - Other invalid UTF-8 input sequences are interpreted as Latin-1 sequences and are
+//   converted into valid UTF-8 instead of being treated as errors.
+template<bool Json>
 static void
-AppendUtf8JsonEscapedWithRoomReserved(
+AppendUtf8WithRoomReserved(
```

Comment:
Note: Make sure output is valid UTF-8. Assume that invalid UTF-8 is actually Latin-1. Take `Json` as a template parameter so that we don't have to write this twice. (I don't want Json to be a dynamic parameter because I don't want an extra condition/branch in the inner loop.)